File: PHP/Algorithm/11_盛最多水的容器.php
<?php

/**
 * 题目
 *     具体题目请查看：https://leetcode-cn.com/problems/container-with-most-water/
 *     图中垂直线代表输入数组 [1,8,6,2,5,4,8,3,7]。在此情况下，容器能够容纳水（表示为蓝色部分）的最大值为 49。
 * 示例 1:
 *     输入: [1,8,6,2,5,4,8,3,7]
 *     输出: 49
 */
class Solution
{
    /**
     * 思路
     *    双指针 (https://leetcode-cn.com/problems/container-with-most-water/solution/sheng-zui-duo-shui-de-rong-qi-by-leetcode/)
     *    双指针，向中间计算面积，获取最大面积
     * 耗时
     *    执行用时 :40 ms, 在所有 PHP 提交中击败了96.81%的用户
     *    内存消耗 :16.2 MB, 在所有 PHP 提交中击败了78.82%的用户
     * 时间复杂度
     *    O（n）
     */
    function maxArea($height)
    {
    	$len      = count($height);
    	$hei      = $len-1; // 高指针
    	$low      = 0;      // 低指针
    	$area     = 0;

    	while ($low != $hei) {
    		
    		$countHeight = $height[$hei] > $height[$low] ? $height[$low] : $height[$hei];
    		$countWide   = $hei - $low;
    		$countArea   = $countHeight * $countWide;

    		// 和已经存在的面积作比较
    		$area = $countArea > $area ? $countArea : $area;

    		if ($height[$hei] >= $height[$low]) {
    			$low++;
    		}
    		else {
    			$hei--;
    		}
    	}

        return $area;
    }
}

$s = new Solution();
var_dump($s->maxArea([1,2,1]));

File: Go/Algorithm/11_盛最多水的容器.go
package main

import (
	"fmt"
)


func main() {
	fmt.Println(maxArea([]int{1,3,2,5,25,24,5}))
}

//思路
//	双指针-在区间的左右两边建立指针,逐渐缩小区间,找到最大的区间
//	https://leetcode-cn.com/problems/container-with-most-water/solution/sheng-zui-duo-shui-de-rong-qi-by-leetcode-solution/
//耗时
//	执行用时： 104 ms , 在所有 Go 提交中击败了 30.89% 的用户
//	内存消耗： 8.3 MB , 在所有 Go 提交中击败了 38.34% 的用户
func maxArea(height []int) int {

	low, hei, res := 0, len(height) - 1, 0

	for hei > low {

		m := min(height[low], height[hei]) * (hei - low)
		if m > res {
			res = m
		}

		if height[low] >= height[hei] {
			hei--
		} else {
			low++
		}
	}

	return res
}

func min(x, y int) int {
	if x > y {
		return y
	}
	return x
}